### PR TITLE
fix upload of assignment1

### PR DESCRIPTION
**Pull request by ellapetter**

### CODE REVIEW

File: Assignment1.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+def q1(a, b):
```
Comment (flerdacodeu) [inline Assignment1.py]: You can use more descriptive names (`q1`, `a`, `b` are not immediately clear) as well as add a docstring describing what the function is meant to do, which argument it takes, and what values it returns.

File: Assignment1.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+def q1(a, b):
+    a = a.lower()
+    b = b.lower()
+    if len(a) != len(b):
+        return False
+    d = {}
+    for letter in a:
+        if letter in d:
```
Comment (flerdacodeu) [inline Assignment1.py]: You can use `dict.get()` to read a value using a default value if not present.
This should allow you to remove the `if`-statement and simplify the code.

File: Assignment1.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            d[letter] += 1
+        else:
+            d[letter] = 1
+    for b_letter in b:
+        if b_letter not in d:
+            return False
+        else:
+            d[b_letter] -= 1
+            if d[b_letter] == 0:
+                d.pop(b_letter)
```
Comment (flerdacodeu) [inline Assignment1.py]: `dict.pop()` is meant to get out the value and remove it.
Since here you do not really care about the value, using `del` would seem more appropriate.

File: Assignment1.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    if d:
+        return False
+    else:
+        return True
+
+
+class Node:
+    def __init__(self, data = None):
+        self.next = None
```
Comment (flerdacodeu) [inline Assignment1.py]: Since `next` and `data` should not be accessed outside the class, the convention is to name them `_next` and `_data` if they are meant to be useful by subclasses, and `__next` and `__data` if they are meant to be private to this class. I would recommend `__next` and `__data`. With these names, they will not actually be accessible to code outside the class. Since you have defined accessors, that's the best practice here.

File: Assignment1.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        self.data = data
+
+    def define_next(self, nextnode):
```
Comment (flerdacodeu) [inline Assignment1.py]: Probably calling this `set_next` would be a better name.
The convention is often to have accessors as setters and getters, e.g., in this case, `set_next()` and `get_next()`.

File: Assignment1.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    def insert(self, node):
+        node.define_next(self.head)
+        self.head = node
+
+
+def q2(lst, k):
+    first_node = lst.head
+    curr_node = first_node
+    if curr_node is not None:
```
Comment (flerdacodeu) [inline Assignment1.py]: This structure can be simplified because the `else`-branch always returns.

If you negate the condition and return, then you can have the current `then`-branch outside the `if`-statement itself.

File: Assignment1.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    def __init__(self, head = None):
+        self.head = head
+
+    def insert(self, node):
+        node.define_next(self.head)
+        self.head = node
+
+
+def q2(lst, k):
+    first_node = lst.head
```
Comment (flerdacodeu) [inline Assignment1.py]: Why not add an accessor for `head` as well?

File: Assignment1.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    curr_node = first_node
+    if curr_node is not None:
+        counter = 1
+    else:   ##the list is empty. so the K'th item is None
```
Comment (flerdacodeu) [inline Assignment1.py]: Do you use `##` for commends, use `#` followed by a space.

File: Assignment1.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        return None
+    while Node.get_next(curr_node) is not None: ##by the end of the loop counter will hold the size of my the given list
```
Comment (flerdacodeu) [inline Assignment1.py]: This comment would be best located on the line before the `while`.
Easier to read there, I think.

File: Assignment1.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        curr_node = Node.get_next(curr_node)
```
Comment (flerdacodeu) [inline Assignment1.py]: You are calling `get_next` twice: in the loop condition and in the loop itself: you can actually move the assignment to the loop condition instead and avoid duplicating this.

Comment (flerdacodeu) [inline Assignment1.py]:
```python
Node.get_next(curr_node)
```

is not the *standard* syntax for calling a method on an object. Use instead:

```python
curr_node.get_next()
```

File: Assignment1.py (hunk 1 above)
Comment (ellapetter) [inline Assignment1.py]: changed, thanks!

File: Assignment1.py (hunk 2 above)
Comment (ellapetter) [inline Assignment1.py]: cool! changed

File: Assignment1.py (hunk 3 above)
Comment (ellapetter) [inline Assignment1.py]: makes sense :)

File: Assignment1.py (hunk 5 above)
Comment (ellapetter) [inline Assignment1.py]: 👍

File: Assignment1.py (hunk 7 above)
Comment (ellapetter) [inline Assignment1.py]: done

File: Assignment1.py (hunk 4 above)
Comment (ellapetter) [inline Assignment1.py]: did not know this! thank you!

File: Assignment1.py (hunk 6 above)
Comment (ellapetter) [inline Assignment1.py]: right, thanks!

File: Assignment1.py (hunk 10 above)
Comment (ellapetter) [inline Assignment1.py]: I'm not sure I understood you correctly - I couldn't find a way to have the assignment inside the while loop itself.. is that what you meant?

**Comment by ellapetter** [comment]
Hi :)
I've uploaded the imporved version of my code + added the tests.
Thanks so much for your review! I replied to one comment with a question, look forward to your response.
Thanks again!

File: Assignment1.py (hunk 10 above)
Comment (flerdacodeu) [inline Assignment1.py]: That's right! Python does not allow you to do that.

But you can move the assignment to before the loop condition is first evaluated, and use an extra variable.

This got me curious, so I put together a micro-benchmark for this.

This is what I got:

```shell
$ python3 performance/run_benchmark.py list_benchmark
option1:
  12.050865086028352 seconds

option2:
  22.274820561055094 seconds

option1 is 1.8x faster
```

See https://github.com/flerdacodeu/codeu/commit/953df0e233d3a8ae762704b441d9b1428bbc2ff4 for the code I used here.

File: Assignment1.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+            d[b_letter] -= 1
+            if d[b_letter] == 0:
+                del d[b_letter]
+    if d:
+        return False
+    else:
+        return True
+
+
+class Node:
```
Comment (flerdacodeu) [inline Assignment1.py]: You can also add a docstring for `Node` and `SinglyLinkedList`, to tell (briefly) what it is meant to do.
Not that for `set_next`, `get_next`, `get_data` it is probably best not to add a docstring, since they are obvious.

File: Assignment1.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        node.define_next(self.head)
+        self.__head = node
+
+    def get_head(self):
+        return self.__head
+
+
+def find_kth_to_last_element(lst, k):
+    '''this function returns the Kth to last element of a list
+    args: linked list lst, int k
```
Comment (flerdacodeu) [inline Assignment1.py]: A useful convention is to have:

```python
    """This function results the Kth to last element of a list.

    Args:
      lst: an instance of SinglyLinkedList, the list from which to extract the element
      k: an integer, the index of the element to extract from the end of the list

    Returns:
      the k-th to last element of the list or None if no such element exists.
```

File: Assignment1.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    first_node = lst.get_head()
+    curr_node = first_node
+    if curr_node is None:
+        return None     # the list is empty. so the K'th item is None
+    counter = 1
+    # by the end of the loop counter will hold the size of my the given list
+    while curr_node.get_next() is not None:
+        curr_node = curr_node.get_next()
+        counter += 1
+    index = counter - k ##calculate the index of the item to be returned
```
Comment (flerdacodeu) [inline Assignment1.py]: Change `##` here as well.

File: Test-Q2.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import unittest
+import Assignment1
+
+
+class Q2Test(unittest.TestCase):
+
+    def test_empty(self):
+        empty_lst = Assignment1.SinglyLinkedList()
+        self.assertEqual(None, Assignment1.find_kth_to_last_element(empty_lst, 2))
```
Comment (flerdacodeu) [inline Test-Q2.py]: You can also add a test for something like a negative index, an index that is beyond the length of the list for a non-empty list.

File: Test-Q2.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+
+class Q2Test(unittest.TestCase):
+
+    def test_empty(self):
+        empty_lst = Assignment1.SinglyLinkedList()
+        self.assertEqual(None, Assignment1.find_kth_to_last_element(empty_lst, 2))
+
+    def test_known_cases(self):
+        # this part will create a list of 1 to 10
+        first = Assignment1.Node(1)
```
Comment (flerdacodeu) [inline Test-Q2.py]: As you add more tests, you might find that creating a list with numbers from 1 to N might be needed in more than one place. For that, you can extract a helper method in the test: as long as the name of the method does not itself start with `test`, it will be ignored by the test runner.

File: Assignment1.py (hunk 11 above)
Comment (ellapetter) [inline Assignment1.py]: Done :)

File: Test-Q2.py (hunk 15 above)
Comment (ellapetter) [inline Test-Q2.py]: Very useful! thank you!
